cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -164,35 +169,43 @@ runQueryTip (AnyConsensusModeParams cModeParams) network mOutFile = do
       consensusMode = consensusModeOnly cModeParams
 
   anyEra <- determineEra cModeParams localNodeConnInfo
-  mEpoch <- mEpochQuery anyEra consensusMode  localNodeConnInfo
   tip <- liftIO $ getLocalChainTip localNodeConnInfo
+
+  let tipSlotNo = case tip of
+        ChainTipAtGenesis -> 0
+        ChainTip slotNo _ _ -> slotNo
+
+  epochResult <- mSlotToEpoch consensusMode localNodeConnInfo tipSlotNo
+    <&> bimap toJsonPastHorizonException tuple3Fst
+
+  let jsonEpoch :: Aeson.Value = either identity toJSON epochResult
+
   let output = encodePretty
         . toObject "era" (Just (toJSON anyEra))
-        . toObject "epoch" mEpoch
+        . toObject "epoch" (Just jsonEpoch)
         $ toJSON tip
   case mOutFile of
     Just (OutputFile fpath) -> liftIO $ LBS.writeFile fpath output
     Nothing                 -> liftIO $ LBS.putStrLn        output
+    
   where
-    mEpochQuery
-      :: AnyCardanoEra
-      -> ConsensusMode mode
+    tuple3Fst :: (a, b, c) -> a
+    tuple3Fst (a, _, _) = a
+
+    mSlotToEpoch
+      :: ConsensusMode mode
       -> LocalNodeConnectInfo mode
-      -> ExceptT ShelleyQueryCmdError IO (Maybe EpochNo)
-    mEpochQuery (AnyCardanoEra era) cMode lNodeConnInfo =
-      case toEraInMode era cMode of
-        Nothing -> return Nothing
-        Just eraInMode ->
-          case cardanoEraStyle era of
-            LegacyByronEra -> return Nothing
-            ShelleyBasedEra sbe -> do
-              let epochQuery = QueryInEra eraInMode $ QueryInShelleyBasedEra sbe QueryEpoch
-              eResult <- liftIO $ queryNodeLocalState lNodeConnInfo Nothing epochQuery
-              case eResult of
-                Left _acqFail -> return Nothing
-                Right eNum -> case eNum of
-                  Left _eraMismatch -> return Nothing
-                  Right epochNum -> return $ Just epochNum
+      -> SlotNo
+      -> ExceptT ShelleyQueryCmdError IO (Either Qry.PastHorizonException (EpochNo, Word64, Word64))
+    mSlotToEpoch cMode lNodeConnInfo slotNo = case cMode of
+      CardanoMode -> do
+        let epochQuery = QueryEraHistory CardanoModeIsMultiEra
+        eResult <- liftIO $ queryNodeLocalState lNodeConnInfo Nothing epochQuery
+        case eResult of
+          Left acqFail -> left (ShelleyQueryCmdAcquireFailure acqFail)
+          Right eraHistory -> return $ slotToEpoch slotNo eraHistory
+
+      mode -> left (ShelleyQueryCmdUnsupportedMode (show mode))
```

Comment:
Rather than showing `ConsensusMode` we should wrap it in `AnyConsensusMode` and propagate that type through the errors. `AnyConsensusMode` can be rendered with `renderMode`

Reply:
Done.  Thanks!